File: ts/src/tools/create-case.ts
import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { FogBugzClient } from "../fogbugz-client.js";

export function registerCreateCase(
  server: McpServer,
  client: FogBugzClient,
) {
  server.tool(
    "create_case",
    "Create a new FogBugz case. At minimum a title is required. Optionally set project, area, milestone, priority, category, tags, assignee, kanban column, parent case, due date, and an initial comment.",
    {
      title: z.string().describe("Title for the new case"),
      project: z.string().optional().describe("Project name"),
      area: z.string().optional().describe("Area name"),
      milestone: z.string().optional().describe("Milestone name"),
      priority: z.string().optional().describe("Priority name (e.g. 'Must Fix', 'Fix If Time')"),
      category: z.string().optional().describe("Category name (e.g. 'Bug', 'Feature')"),
      tags: z
        .array(z.string())
        .optional()
        .describe("Tags to set on the case"),
      assigned_to: z.string().optional().describe("Full name of the person to assign the case to"),
      kanban_column: z.string().optional().describe("Kanban column name"),
      parent_case: z.number().optional().describe("Parent case number"),
      due_date: z.string().optional().describe("Due date in ISO 8601 format (e.g. '2025-12-31T00:00:00Z')"),
      comment: z.string().optional().describe("Initial comment / description for the case"),
    },
    { destructiveHint: true },
    async (params) => {
      const fields: Record<string, unknown> = {
        sTitle: params.title,
      };

      if (params.project !== undefined) fields.sProject = params.project;
      if (params.area !== undefined) fields.sArea = params.area;
      if (params.milestone !== undefined) fields.sFixFor = params.milestone;
      if (params.priority !== undefined) fields.sPriority = params.priority;
      if (params.category !== undefined) fields.sCategory = params.category;
      if (params.tags !== undefined) fields.sTags = params.tags.join(",");
      if (params.assigned_to !== undefined)
        fields.sPersonAssignedTo = params.assigned_to;
      if (params.kanban_column !== undefined)
        fields.sKanbanColumn = params.kanban_column;
      if (params.parent_case !== undefined)
        fields.ixBugParent = params.parent_case;
      if (params.due_date !== undefined) fields.dtDue = params.due_date;
      if (params.comment !== undefined) fields.sEvent = params.comment;

      const result = await client.newCase(fields);
      const caseUrl = client.getCaseUrl(result.case.ixBug);

      return {
        content: [
          {
            type: "text",
            text: `Case ${result.case.ixBug} created successfully.\nTitle: ${params.title}\nLink: ${caseUrl}`,
          },
        ],
      };
    },
  );
}

File: python/src/fogbugz_mcp/tools/create_case.py
from __future__ import annotations

from typing import TYPE_CHECKING

from mcp.server.fastmcp import FastMCP

if TYPE_CHECKING:
    from ..fogbugz_client import FogBugzClient


def register_create_case(mcp: FastMCP, client: FogBugzClient) -> None:
    @mcp.tool(
        name="create_case",
        description=(
            "Create a new FogBugz case. At minimum a title is required. Optionally set "
            "project, area, milestone, priority, category, tags, assignee, kanban column, "
            "parent case, due date, and an initial comment."
        ),
        annotations={"destructiveHint": True},
    )
    async def create_case(
        title: str,
        project: str | None = None,
        area: str | None = None,
        milestone: str | None = None,
        priority: str | None = None,
        category: str | None = None,
        tags: list[str] | None = None,
        assigned_to: str | None = None,
        kanban_column: str | None = None,
        parent_case: int | None = None,
        due_date: str | None = None,
        comment: str | None = None,
    ) -> str:
        """Create a new FogBugz case."""
        fields: dict = {"sTitle": title}
        if project is not None:
            fields["sProject"] = project
        if area is not None:
            fields["sArea"] = area
        if milestone is not None:
            fields["sFixFor"] = milestone
        if priority is not None:
            fields["sPriority"] = priority
        if category is not None:
            fields["sCategory"] = category
        if tags is not None:
            fields["sTags"] = ",".join(tags)
        if assigned_to is not None:
            fields["sPersonAssignedTo"] = assigned_to
        if kanban_column is not None:
            fields["sKanbanColumn"] = kanban_column
        if parent_case is not None:
            fields["ixBugParent"] = parent_case
        if due_date is not None:
            fields["dtDue"] = due_date
        if comment is not None:
            fields["sEvent"] = comment

        result = await client.new_case(fields)
        case_url = client.get_case_url(result["case"]["ixBug"])

        return f"Case {result['case']['ixBug']} created successfully.\nTitle: {title}\nLink: {case_url}"
